File: plugins/scrape/src/main/java/no/lillehaug/landingsopplysninger/scrape/parse/Parser.java
package no.lillehaug.landingsopplysninger.scrape.parse;

import no.lillehaug.landingsopplysninger.api.Leveringslinje;
import org.apache.http.HttpResponse;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.concurrent.FutureCallback;
import org.apache.http.impl.nio.client.CloseableHttpAsyncClient;
import org.apache.http.message.BasicNameValuePair;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import reactor.core.publisher.Flux;
import reactor.core.publisher.FluxSink;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.Charset;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeFormatterBuilder;
import java.util.Arrays;
import java.util.List;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class Parser {
    private static final Logger logger = LoggerFactory.getLogger(Parser.class);
    private static final Pattern vesselPattern = Pattern.compile("[^(]*\\((.*)\\)");
    public static final Charset latin = Charset.forName("ISO-8859-1");

    private final String url;

    public Parser(String url) {
        this.url = url;
    }

    public Flux<Leveringslinje> fetchAndParseForRegistration(String registration, CloseableHttpAsyncClient client) {
        return Flux.create(sink -> {
            doPost(registration, client, sink);
        });
    }

    private void doPost(String registration, CloseableHttpAsyncClient client, FluxSink<Leveringslinje> sink) {
        HttpPost post = new HttpPost(url);
        post.setEntity(new UrlEncodedFormEntity(Arrays.asList(
                new BasicNameValuePair("p_arg_names", "p_regmerke"),
                new BasicNameValuePair("p_arg_values", registration)
        ), latin));
        FutureCallback<HttpResponse> callback = new FutureCallback<HttpResponse>() {
            @Override
            public void completed(HttpResponse result) {
                try {
                    String html = read(result.getEntity().getContent());
                    parseHtml(html, sink);
                    sink.complete();
                } catch (IOException e) {
                    logger.error("Error reading response", e);
                    sink.error(e);
                }
            }

            @Override
            public void failed(Exception ex) {
                logger.error("Request failed", ex);
                sink.error(ex);
            }

            @Override
            public void cancelled() {
                sink.complete();
            }
        };
        client.execute(post, callback);
    }

    private void parseHtml(String html, FluxSink<Leveringslinje> sink) {
        Document document = Jsoup.parse(html);
        Elements tables = document.getElementsByTag("table");
        Elements lines = tables.select("tr");

        String previousFartøy = "Ukjent";
        String previousLandingsdato = "Ukjent";
        String previousMottak= "Ukjent";

        for (Element line : lines) {
            Elements columns = line.select("td");
            if (columns.size() > 0) {

                String fartøy = columns.get(0).html();
                String landingsdato = columns.get(1).html();
                String mottak = columns.get(2).html();
                String fiskeslag = columns.get(3).html();
                String tilstand = columns.get(4).html();
                String størrelse = columns.get(5).html();
                String kvalitet = columns.get(6).html();
                String nettovekt = columns.get(7).html();
                Matcher matcher = vesselPattern.matcher(fartøy);
                String fartøyKjennemerke = matcher.matches() ? matcher.group(1) : fartøy;

                previousFartøy = defaultOrValue(previousFartøy, fartøyKjennemerke);
                previousLandingsdato = defaultOrValue(previousLandingsdato, landingsdato);
                previousMottak = defaultOrValue(previousMottak, mottak);

                sink.next(new Leveringslinje(
                                             previousFartøy,
                                             tryParseDate(previousLandingsdato),
                                             previousMottak,
                                             fiskeslag,
                                             tilstand,
                                             størrelse,
                                             kvalitet,
                                             Double.valueOf(nettovekt)));
            }
        }


    }

    private static final List<DateTimeFormatter> dataParsers = Arrays.asList(
            new DateTimeFormatterBuilder()
                    .parseCaseInsensitive()
                    .appendPattern("dd-MMM-yy")
                    .toFormatter(),
            new DateTimeFormatterBuilder()
                    .parseCaseInsensitive()
                    .appendPattern("dd-MMM-yy")
                    .toFormatter(Locale.forLanguageTag("NO")));
    private LocalDate tryParseDate(String landingsDato) {

        for (DateTimeFormatter parser: dataParsers) {
            try {
                return LocalDate.parse(landingsDato, parser);
            } catch(Exception e) {
                // try next
            }
        }
        throw new IllegalStateException("No parsers for ${landingsDato}");
    }

    private String defaultOrValue(String defaultValue, String value){
        if (value == null || value.trim().isEmpty() || value.equals("&nbsp;")) {
            return defaultValue;
        } else {
            return value;
        }
    }

    public static String read(InputStream input) throws IOException {
        try (BufferedReader buffer = new BufferedReader(new InputStreamReader(input, latin))) {
            return buffer.lines().collect(Collectors.joining("\n"));
        }
    }
}

File: plugins/scrape/src/main/java/no/lillehaug/landingsopplysninger/scrape/parse/Parser.kt
package no.lillehaug.landingsopplysninger.scrape.parse

import no.lillehaug.landingsopplysninger.api.Leveringslinje
import org.apache.commons.io.IOUtils
import org.apache.http.client.entity.UrlEncodedFormEntity
import org.apache.http.client.methods.HttpGet
import org.apache.http.client.methods.HttpPost
import org.apache.http.impl.client.CloseableHttpClient
import org.apache.http.message.BasicNameValuePair
import org.jsoup.Jsoup
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeFormatterBuilder
import java.util.*
import javax.script.ScriptEngineManager

val jsPattern = ".*innerHTML = \"(.*)\";.*".toRegex(RegexOption.MULTILINE).toPattern()
val tagPattern = "\\</?[^>]*\\>".toRegex().toPattern()
val engine = ScriptEngineManager().getEngineByName("nashorn")

class Parser (private val url: String){

    fun fetchAndParseForRegistration(registration: String, httpclient: CloseableHttpClient) : List<Leveringslinje> {

        val get = httpclient.execute(HttpGet(url)).use {
            val html = IOUtils.toString(it.entity.content, "ISO-8859-1")
            parseForm(html)
        }

        if(get == null) emptyList<Leveringslinje>()

        val entity = UrlEncodedFormEntity(
                listOf(
                        BasicNameValuePair("regmerke", registration),
                        BasicNameValuePair("svar", get?.second)
                ))

        val action = get?.first
        val httpPost = HttpPost("http://www.rafisklaget.no${action}")
        httpPost.entity = entity

        return httpclient.execute(httpPost).use {
            val html = IOUtils.toString(it.entity.content, "ISO-8859-1")
            parseFromHtml(html, registration)
        }
    }

    private fun parseForm(html: String): Pair<String, String>? {
        val parsed = Jsoup.parse(html)
        val form = parsed.getElementById("reportfilter")
        val action = form.attr("action")

        val elementsByTag = parsed.getElementsByTag("script")
        val script = elementsByTag[0]
        val text = script.data().split("\n")
                .firstOrNull { it.contains("innerHTML = ") }
        if(!text.isNullOrBlank()) {
            val matcher = jsPattern.matcher(text)
            if(matcher.matches()) {
                val expression = matcher.group(1)
                val stripped = tagPattern.matcher(expression)
                        .replaceAll("")
                        .replace("=", "")
                val eval = engine.eval(stripped)

                return Pair(action, eval.toString())
            }
        }
        return null
    }

    fun parseFromHtml(html: String, fartøy: String)  : List<Leveringslinje> {
        val document = Jsoup.parse(html)
        val tables = document.getElementsByTag("table")
        val lines = tables.select("tr")

        val entries = mutableListOf<Entry>()
        var previousLandingsdato = "Ukjent"
        var previousMottak = "Ukjent"

        for (line in lines) {
            val columns = line.select("td")
            if (columns.size > 0) {
                val landingsdato = columns[0].html()
                val mottak = columns[1].html()
                val fiskeslag = columns[2].html()
                val tilstand = columns[3].html()
                val størrelse = columns[4].html()
                val kvalitet = columns[5].html()
                val nettovekt = columns[6].html()

                previousLandingsdato = defaultOrValue(previousLandingsdato, landingsdato)
                previousMottak = defaultOrValue(previousMottak, mottak)
                entries.add(Entry(fartøy, previousLandingsdato, previousMottak, fiskeslag, tilstand, størrelse, kvalitet, nettovekt))
            }
        }

        return entries.map {
            Leveringslinje(
                    it.fartøy,
                    tryParseDate(it),
                    it.mottak,
                    it.fiskeslag,
                    it.tilstand,
                    it.størrelse,
                    it.kvalitet,
                    it.nettovekt.toDouble()
            )
        }
    }

    private val dateParsers = listOf(
            DateTimeFormatter.ofPattern("dd.MM.yyyy"),
            DateTimeFormatterBuilder()
                    .parseCaseInsensitive()
                    .appendPattern("dd-MMM-yy")
                    .toFormatter(),
            DateTimeFormatterBuilder()
                    .parseCaseInsensitive()
                    .appendPattern("dd-MMM-yy")
                    .toFormatter(Locale.forLanguageTag("NO")))
    private fun tryParseDate(it: Entry) : LocalDate {

        val landingsDato = it.landingsdato
        for (parser in dateParsers) {
            try {
                return LocalDate.parse(landingsDato, parser)
            } catch(e: Exception) {
                // lol
            }
        }
        throw IllegalStateException("No parsers for ${landingsDato}")
    }

    private fun defaultOrValue(defaultValue: String, value: String? ) : String {
        return if (value == null || value.isBlank() || value == "&nbsp;") {
            defaultValue
        } else {
            value
        }
    }

    data class Entry(val fartøy: String, val landingsdato: String, val mottak: String, val fiskeslag: String, val tilstand: String, val størrelse: String, val kvalitet: String, val nettovekt: String)
}
